### ISSUE TITLE
Very slow cloning of "file://" repository format

### ISSUE THREAD

**Issue by torvalds** [libgit2/libgit2#2836]
So we use libgit2 in subsurface as one of the ways to access the dive data, and am starting to add some support for actually accessing repositories remotely.

While doing that, for debugging I don't want to actually have to go over the network (because I'm in remote locations and on planes etc right now), and so I was using my local dive repository at "file:///home/torvalds/.." as the "remote" repo.

It is *incredibly* slow. Regular git clones it in a couple of seconds (doing a true local clone with just hardlinks is obviously faster still, but I actually want this to act like a remote clone). But using "git_clone()" from libgit2 (just plain F21 version, so it's 0.21 rather than current HEAD) takes *minutes*. 

I have no idea what the heck would make it that slow. But it's orders of magnitude slower than real git. 

I can fake things, and just use "/home/torvalds/..." instead of the "file://" format, and things work ok, but I'm wondering if this is perhaps some bug in libgit2 that people just hadn't realized existed, because most people never use the "file://" form.

**Comment by JIghtuse** [comment]
I dig into sources with debugger and simple clone utility. Looks like it is proper default behavior. User can set a field `local` in `git_clone_options` struct for needed behavior:
```c
/**
 * Options for bypassing the git-aware transport on clone. Bypassing
 * it means that instead of a fetch, libgit2 will copy the object
 * database directory instead of figuring out what it needs, which is
 * faster. If possible, it will hardlink the files to save space.
 */
typedef enum {
	/**
	 * Auto-detect (default), libgit2 will bypass the git-aware
	 * transport for local paths, but use a normal fetch for
	 * `file://` urls.
	 */
	GIT_CLONE_LOCAL_AUTO,
	/**
	 * Bypass the git-aware transport even for a `file://` url.
	 */
	GIT_CLONE_LOCAL,
	/**
	 * Do no bypass the git-aware transport
	 */
	GIT_CLONE_NO_LOCAL,
	/**
	 * Bypass the git-aware transport, but do not try to use
	 * hardlinks.
	 */
	GIT_CLONE_LOCAL_NO_LINKS,
} git_clone_local_t;

```
So, if you have `file://` format and you haven't set `clone_opts.local`, by default it would not use "local clone". Setting this option to GIT_CLONE_LOCAL makes cloning as fast as in `/home/` format, setting it to GIT_CLONE_NO_LOCAL makes it slow in both cases.

**Comment by torvalds** [comment]
On Jan 24, 2015 7:30 AM, "Boris Egorov" <notifications@github.com> wrote:
>
> I dig into sources with debugger and simple clone utility. Looks like it
is proper default behavior. User can set a field local in git_clone_options
struct for needed behavior:

You don't understand. I don't *want* a local clone that just links directly
too the files etc. Io actually want the while new repacking etc, so that
things look as much like a real remote clone as possible

And that us quite fast with real git.

But libgit2 makes that very very slow. Much slower than it should be.
Incredibly slow. Native git takes three seconds on my laptop (the local
"link directly too files" is basically instantaneous). Libgit2 takes
forever. As in minutes.

*That* is the problem. The horrible performance issues.

      Linus

**Comment by JIghtuse** [comment]
Oh, ok. Now I got it.

I am not too familiar with libgit2 codebase, but I've run cloning of
local repo and attach process to 'strace -c'. It probably can help
devs to find some clues what's going on.

```txt
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 29.97    0.002785           0    156146           ftruncate
 14.74    0.001370           0    253589           munmap
 12.62    0.001173           0    253623           mmap
 10.98    0.001020           1      1888           mkdir
 10.93    0.001016           1       791           brk
 10.02    0.000931           0      3430        14 open
  3.69    0.000343           0    198474    198456 access
  3.40    0.000316           0     90043      8406 lstat
  2.73    0.000254           0     76973     47205 stat
  0.36    0.000033           0     11536           close
  0.30    0.000028           0     16240           getdents
  0.13    0.000012           0       146           read
  0.13    0.000012           0      3521           write
  0.00    0.000000           0        30           fstat
```

**Comment by minshallj** [comment]
Maybe this will help also... When doing a simple clone with `strace`
```
access("/tmp/test-repo/.git/objects/23/a34618ec87f57deca15a0266a3016410d876a2", F_OK) = -1 ENOENT (No such file or directory)
```
Is this an old (or new) way of storing objects(`.git/objects/...`)? Some repos I have use this style and some don't... When I do a simple clone of libgit2 there are 254082 lines grepped that match "No such file"

I've been able to trace this to the line https://github.com/libgit2/libgit2/blob/master/src/fetch.c#L137
but I haven't found where the callback came from... yet...

**Comment by ethomson** [comment]
@torvalds - indeed, this is very unfortunate.  I think that your analysis is correct - nobody's yet been inclined to do a local clone with the non-local mechanism since it's a little bit more exotic.  I've got a few unrelated things cooking at the moment (plus $DAYJOB), but I'll try to take a look in the near future.  Thanks for the report.

To answer some of the other questions / comments that have popped up on this thread:

@jlghtuse - by default, clone operates in [`GIT_CLONE_LOCAL_AUTO`](https://github.com/libgit2/libgit2/blob/master/include/git2/clone.h#L39) mode, which tries to figure out how it should deal with the clone in a manner somewhat similar to `git-clone(1)`.  By default, it will hardlink the objects (or, on braindead systems, copy) from the origin repository when a local path is specified, but treat `file://` URLs more like they were actual remotes.

@minshallj - this means that using a `file://` URL (or specifically using `GIT_CLONE_NO_LOCAL` mode) will give you a single packfile in the new repository, while using `GIT_CLONE_LOCAL` mode will give you whatever the origin had, which may be a collection of packfiles and/or loose objects.

**Comment by carlosmn** [comment]
Our version of the logic to feed pack-objects which objects to use is somewhat naïve but taking minutes is surprising. Cloning libgit2 (which results in a packfile of 16MB) with itself takes about 30-40s with both with master and maint/v0.21 and most of the time is spent creating the deltas (according to perf).

Does subsurface use very shallow or very deep trees to store the data? That could affect how good our shortcuts are. Is there some subsurface-created repo publicly avaliable which we could use to test?

**Comment by peff** [comment]
If your source repository is fully packed, then you should not need to do any delta search. Is libgit2 missing the optimization to avoid looking for deltas between objects that are already in the same pack?

**Comment by carlosmn** [comment]
We are indeed missing that optimisation, partly due to modularity. Our odb interface does not present whether something is packed or as a delta, we can only ask for full objects. This is obviously less than ideal, but since the main use-case for our packing is to push a few objects from a developer's repo, it's generally not a big deal in practice.

**Comment by peff** [comment]
Yeah, your `pack-objects` implementation can never compete with git-core's at that level of abstraction. Git tries very hard to make serving a fetch as light as possible on the server-side both by reusing deltas, but also by reusing the on-disk packed data (so we don't even have to re-deflate).

I wonder how hard it would be to convince libgit2 to just call `git-upload-pack` to do the server side of a local `file://` clone (and fall back to using its own slower implementation if the user doesn't have regular git installed).

**Comment by ethomson** [comment]
Not particularly hard; one could plug in a custom transport that did this.  The hard part would be locating `git` and dealing with `popen` oddities across platforms.  I think this would be generally beneficial.

It's not obvious to me the best way to get this to consumers.  I wouldn't want to see this in the library itself, but maybe a `contrib` folder at the root that contains these sorts of things?  The risk here is that it's easy for those sorts of things to bit rot or to become a maintenance headache (or both), and that it's not always obvious to consumers to go looking in there for what they want or might expect in the box.

**Comment by carlosmn** [comment]
Calling out to other programs is something We Don't Do™ in libgit2, but it's not like we *have* to stay at this level of abstraction. My plan for solving this is to simply have a couple of optional functions in the interface which a backend can fill to say that it can provide us with already-packed or already-deltified data. If you're implementing your own backend, then you can just not fill it and be slower, while the default one can give us the data (but then someone has to have this as their next item in things to fix).

But I have a hunch that the packing itself is what's slowing down subsurface, but rather the logic to create the list of objects to fill. In the push case we have some short-circuiting logic based on diffs, but the local transport still wants to insert the same objects into the map over and over again. I started writing some logic to handle this more like git does (and we now do with the revwalker), but it has a ways to go.

**Comment by carlosmn** [comment]
With #2972  and #2973 which provide an algorithmic improvement and multiple threads when using the local transport (thus emulating what we'd get by calling upload-pack), the performance has improved considerably.

There's still quite some road ahead, but it should be quite noticeable already.

**Comment by PaulDance** [comment]
@ethomson maybe?
